[Cloud Security] Rules page enhancements

JordanSh opened the pull request:
## Summary

Added Rule numbers for the rules table, as well as section filter. the section filter is responsive with max width of 300px, and has keyboard support. when empty displays all sections.

https://github.com/elastic/kibana/assets/51442161/b068813c-195f-4b4d-8be5-da29ad620366

kfirpeled commented:
@JordanSh can we sort the values alphabetically? both for rule number and CIS section (ignore case sensitive)

JordanSh commented:
@kfirpeled we don't actually support on column click sorting for this table, the sorting is done on the BE and it's using a default value from the schema. I'm not sure why since we only have about a hundred rules per benchmark so we probably just fetch all and support in memory sorting and pagination. 

This change is a bit a bigger and we can probably add it to the iterations list for next time. I'll try to change the default sorting to the rule number for now.

on another note @tinnytintin10, i removed the option of opening the flyout on row click, it makes copying the data from the table to the search bars extremely annoying. let me know if you want me to revert it back. here is the before & after video:

https://github.com/elastic/kibana/assets/51442161/d1691363-4ff2-4a88-aef9-76e45f8bdda7

JordanSh commented:
@kfirpeled 

I've added a custom sorting function to the data, we need a custom one since the rule numbers are not versions and do not always support semantic versioning. for example instead of rule number `2.0.0` we have `2.0`.
Editing mappings or adding runtime fields ended up being complicated and I think they were not worth the time investment at this stage.

I've used `semver.compare` for most of the numbers and local compare for the rest. I've added a bunch of tests for edge cases, missing values, empty arrays, single data arrays, and different kind of number format. For now it seems to function well, but it something to keep an eye on and maybe consider putting more work on outside of the "quick" wins.

maxcold commented on x-pack/plugins/cloud_security_posture/common/utils/helpers.ts:
```diff
@@ -46,8 +47,12 @@ export const extractErrorMessage = (e: unknown, defaultMessage = 'Unknown Error'
 
-export const getBenchmarkTypeFilter = (type: BenchmarkId): string =>
-  `${CSP_RULE_TEMPLATE_SAVED_OBJECT_TYPE}.attributes.metadata.benchmark.id: "${type}"`;
+export const getBenchmarkTypeFilter = (type: BenchmarkId, section?: RuleSection): string =>
```
nit: it's not only type filter anymore, maybe just `getBenchmarkFilter`

maxcold commented on x-pack/plugins/cloud_security_posture/public/pages/rules/rules_container.tsx:
```diff
@@ -77,12 +79,31 @@ export const RulesContainer = () => {
     [data, error, status, rulesQuery]
   );
 
+  // We need to make this call again without the filters. this way the section list is always full
```
the cleanest would be to either have a route on the server that returns the sections or do all the filtering/sorting on the client, but I guess we don't want to invest much into the rules page. I hope it won't bite us in the future :)

maxcold commented on x-pack/plugins/cloud_security_posture/server/routes/csp_rule_template/get_csp_rule_templates.test.ts:
```diff
@@ -0,0 +1,89 @@
+import { getSortedCspRulesTemplates } from './get_csp_rule_template';
+import { CspRuleTemplate } from '../../../common/schemas';
+
+describe('getSortedCspRulesTemplates', () => {
```
👍

maxcold commented on x-pack/plugins/cloud_security_posture/server/routes/csp_rule_template/get_csp_rule_template.ts:
```diff
@@ -23,6 +25,21 @@ import {
+    if (versionA !== null && versionB !== null) {
+      return semverCompare(versionA, versionB);
+    } else {
+      return String(a?.metadata?.benchmark?.rule_number).localeCompare(
```
nit: move `a?.metadata?.benchmark?.rule_number` and `b?.metadata?.benchmark?.rule_number` to a const

maxcold reviewed (approved):
tested, works as expected. Couple of nit comments
